cache/merge.go

```diff
@@ -0,0 +1,93 @@
+package cache
+
+import (
+	"errors"
+	"github.com/spacemeshos/merkle-tree"
+	"io"
+)
+
+func Merge(caches []CacheReader) (*Reader, error) {
+	if len(caches) < 2 {
+		return nil, errors.New("number of caches must be at least 2")
+	}
+
+	// Aggregate caches' layers by height.
+	layerGroups := make(map[uint][]LayerReadWriter)
+	for _, cache := range caches {
+		for height, layer := range cache.Layers() {
+			layerGroups[height] = append(layerGroups[height], layer)
+		}
+	}
+
+	// Group layer groups.
+	layers := make(map[uint]LayerReadWriter)
+	for height, layerGroup := range layerGroups {
+		if len(layerGroup) != len(caches) {
+			return nil, errors.New("number of layers per height mismatch")
+		}
+
+		group, err := groupLayers(layerGroup)
+		if err != nil {
+			return nil, err
+		}
+		layers[height] = group
+	}
+
+	cache := &cache{layers: layers}
+	return &Reader{cache}, nil
+}
+
+func BuildTop(cacheReader CacheReader) (*Reader, []byte, error) {
+	// Find the cache highest layer.
+	var maxHeight uint
+	for height := range cacheReader.Layers() {
+		if height > maxHeight {
+			maxHeight = height
+		}
+	}
+
+	// Create a new subtree with the cache highest layer as its leaves.
+	subtreeWriter := NewWriter(MinHeightPolicy(0), MakeSliceReadWriterFactory())
```

Comment:
The cache's caching policy should be honored (e.g. it might want to cache on-disk and not in-memory, or not cache all layers).

To do this, you can wrap the existing policies such that they receive the adjusted layer height. Basically create a closure that adds `maxHeight` to `height` and calls the internal function.